modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosControllerEpics.scala

```diff
@@ -69,78 +61,117 @@ class GmosControllerEpics[T<:GmosController.SiteDependentTypes](encoders: GmosCo
     case Order.TWO  => 2
   }
 
-  implicit val disperserLambdaEncoder: EncodeEpicsValue[Length, Double] = EncodeEpicsValue((l: Length) => l.toNanometers)
+  implicit val disperserLambdaEncoder: EncodeEpicsValue[Length, Double] =
+    EncodeEpicsValue((l: Length) => l.toNanometers)
 
-  implicit val useElectronicOffsetEncoder: EncodeEpicsValue[UseElectronicOffset, Int] = EncodeEpicsValue(_.allow.fold(1, 0))
+  implicit val useElectronicOffsetEncoder: EncodeEpicsValue[UseElectronicOffset, Int] =
+    EncodeEpicsValue(_.allow.fold(1, 0))
 
-  private def setShutterState(s: ShutterState): SeqAction[Unit] = s match {
-    case UnsetShutter => SeqAction.void
-    case sh           => DC.setShutterState(encode(sh))
+  private def setShutterState(s: ShutterState): List[SeqAction[Unit]] = s match {
+    case UnsetShutter => List.empty
+    case sh           => {
+      val encodedVal = encode(sh)
+      smartSetParam(encodedVal, GmosEpics.instance.shutterState, DC.setShutterState(encode(sh)))
+    }
   }
 
   private def roiNumUsed(s: RegionsOfInterest): Int = s.rois match {
     case Right(rois) => rois.length
     case Left(_)     => 1
   }
 
-  private def setROI(binning: CCDBinning, s: RegionsOfInterest): SeqAction[Unit] = s.rois match {
-    case Left(b)     =>
-      roiParameters(binning, 1, encoders.builtInROI.encode(b))
-    case Right(rois) =>
-      rois.zipWithIndex.map { case (roi, i) =>
+  private def setROI(binning: CCDBinning, s: RegionsOfInterest): List[SeqAction[Unit]] =
+    s.rois match {
+      case Left(b)     => roiParameters(binning, 1, encoders.builtInROI.encode(b))
+      case Right(rois) => rois.zipWithIndex.flatMap { case (roi, i) =>
         roiParameters(binning, i + 1, ROIValues.fromOCS(roi))
-      }.sequence.flatMap(_ => SeqAction.void)
+      }
+    }
+
+  private def roiParameters(binning: CCDBinning, index: Int, roi: Option[ROIValues])
+  : List[SeqAction[Unit]] = (roi, DC.rois.get(index)).mapN { (roi, r) =>
+    r.setCcdXstart1(roi.xStart.value) *>
+    r.setCcdXsize1(roi.xSize.value / binning.x.getValue) *>
+    r.setCcdYstart1(roi.yStart.value) *>
```

Comment:
I've never known what `*>` was used for before. This was very helpful.

Reply:
It's a `flatMap` that ignores the input.

Reply:
Thanks for the explanation. That will be useful in the future for me, I'm sure. Looks good!